v2/pkg/engine/datasource/graphql_datasource/graphql_tws_handler.go

```diff
@@ -1,88 +1,166 @@
 package graphql_datasource
 
 import (
+	"bufio"
 	"context"
 	"encoding/json"
 	"errors"
 	"fmt"
-	"github.com/wundergraph/graphql-go-tools/v2/pkg/engine/resolve"
-	"io"
 	"net"
-	"strconv"
+	"strings"
 	"time"
 
+	"github.com/gobwas/ws/wsutil"
+	"github.com/wundergraph/graphql-go-tools/v2/pkg/engine/resolve"
+
 	"github.com/buger/jsonparser"
 	log "github.com/jensneuse/abstractlogger"
-	"nhooyr.io/websocket"
 )
 
 // gqlTWSConnectionHandler is responsible for handling a connection to an origin
 // it is responsible for managing all subscriptions using the underlying WebSocket connection
 // if all Subscriptions are complete or cancelled/unsubscribed the handler will terminate
 type gqlTWSConnectionHandler struct {
-	conn               *websocket.Conn
-	ctx                context.Context
-	log                log.Logger
-	subscribeCh        chan Subscription
-	nextSubscriptionID int
-	subscriptions      map[string]Subscription
-	readTimeout        time.Duration
+	conn                          net.Conn
+	requestContext, engineContext context.Context
+	log                           log.Logger
+	options                       GraphQLSubscriptionOptions
+	updater                       resolve.SubscriptionUpdater
+}
+
+func (h *gqlTWSConnectionHandler) ServerClose() {
+	h.updater.Done()
+	_ = h.conn.Close()
+}
+
+func (h *gqlTWSConnectionHandler) ClientClose() {
+	h.updater.Done()
+	req := fmt.Sprintf(completeMessage, "1")
+	err := wsutil.WriteClientText(h.conn, []byte(req))
+	if err != nil {
+		h.log.Error("failed to write complete message", log.Error(err))
+	}
+	_ = h.conn.Close()
+}
+
+func (h *gqlTWSConnectionHandler) Subscribe() error {
+	return h.subscribe()
+}
+
+func (h *gqlTWSConnectionHandler) ReadMessage() (done, timeout bool) {
```

Comment:
Heartbeat handling for subscriptions over multipart is currently not handled when using epoll. We also need to ensure that this logic is part of epoll to avoid excessive use of goroutines.

See https://github.com/wundergraph/graphql-go-tools/blob/7b40eb8fc88127ce992b7fb6b71a48b963546f65/v2/pkg/engine/datasource/graphql_datasource/graphql_tws_handler.go#L149

Reply:
Heartbeat doesn't even belong into the origin client implementation. It's a concern of the resolver, which is router client facing. I'll give a longer explanation on slack, but the short one is that the heartbeat implementation doesn't belong into the origin client, because all it does it call updater.Heartbeat, which the resolver can do by itself, which means that we don't even need to expose heartbeat to the updater interface.